en/ros/ros2_comm.md

```diff
@@ -206,6 +217,28 @@ E.g. in the case of `sensor_combined` use `ros2 topic hz /fmu/sensor_combined/ou
    ```
 
 
+### Custom uORB/ROS2 Topics
+
+ROS 2 needs to have the _same_ message definitions that were used to create the XRCE-DDS client module in the PX4 Firmware.
+These are automatically exported to [PX4/px4_msgs](https://github.com/PX4/px4_msgs) for main and release builds, and you can just clone them into your workspace.
+
+If working on a custom build that has other messages, you would:
+- update the [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml) to add needed topics to your PX4 build.
+- Copy the msg definitions out of PX4 source tree into any ROS2 workspace that needs them.
+  The messages must be in a folder named `px4_msgs/msg/`.
+  
+  :::note
+  ROS2 messages are sent with the topic and type name defined the yaml file, but do not include the message definition itself (which is why these have to be supplied to the ROS2 workspace, and have to match the message that is sent).
+  For example, the following yaml entry defines the `px4_msgs::msg::VehicleOdometry` type.
+  ```yaml
+  - topic: /fmu/out/vehicle_odometry
+    type: px4_msgs::msg::VehicleOdometry
+  ```
+
+  The ROS toolchain uses the prefix in the message type (i.e. `px4_msgs::msg::`) to determine the search path for message definitions.
+  :::
+
+
 ## ROS 2 Example Applications
```

Comment:
@beniaminopozzan @Jaeyoung-Lim Well, I can run run @Jaeyoung-Lim [example](https://github.com/Jaeyoung-Lim/px4-offboard/) with the XRCE-DDS agent, which is nice. Unfortunately it is not quite as good as the examples currently used by this document, which are the advertiser and listener examples in px4_ros_com https://github.com/PX4/px4_ros_com/tree/main/src/examples - in the sense that these show data in/out of ROS2/PX4 - ie. both main use cases.

The problem is that I can't get these to work with XRCE-DDS agent. Any ideas on how they might be modified to work? I don't actually understand ROS and the launch behaviour :-(. 








https://github.com/PX4/px4_ros_com/tree/main/src/examples/advertisers

Reply:
@hamishwillee for the the listener examples the issue is about the ROS topic quality of service which does not matches between publishers and subscribers. This PR https://github.com/PX4/px4_ros_com/pull/174 fixes it for them.

This is actually a good point to bring up: ROS2 default QoS https://docs.ros.org/en/foxy/Concepts/About-Quality-of-Service-Settings.html#qos-profiles is different from the one used by PX4, therefore every time a user wants to create a subscriber it need to specify the right QoS otherwise it won't be able to listen to the topics published by PX4.
Relevant merged PR about this is https://github.com/PX4/PX4-Autopilot/pull/20624

Citing ROS2 documentation:

> By default, publishers and subscriptions in ROS 2 have “keep last” for history with a queue size of 10, “reliable” for reliability, “volatile” for durability, and “system default” for liveliness. Deadline, lifespan, and lease durations are also all set to “default”.

PX4 instead uses
```cpp
uxrQoS_t qos = {
  .durability = UXR_DURABILITY_TRANSIENT_LOCAL,
  .reliability = UXR_RELIABILITY_BEST_EFFORT,
  .history = UXR_HISTORY_KEEP_LAST,
  .depth = 0,
};
```
for publishers, and
```c++
uxrQoS_t qos = {
  .durability = UXR_DURABILITY_VOLATILE,
  .reliability = UXR_RELIABILITY_BEST_EFFORT,
  .history = UXR_HISTORY_KEEP_LAST,
  .depth = queue_depth,
};
```
for subscribers, which corresponds to the ROS2 predefined QoS _sensor data_ (the one that the PR chooses). Not all combinations publisher-subscriber are possible: https://docs.ros.org/en/foxy/Concepts/About-Quality-of-Service-Settings.html#qos-compatibilities .